### PR TITLE
chore(deps): upgrade Express 4 -> 5

### PR DESCRIPTION
**Pull request by adriangohjw**
## Summary

- Bumps `express` (`^4.22.2` -> `^5.2.1`), `@types/express` (`^4.17.25` -> `^5.0.6`), and `express-joi-validation` (`^5.0.1` -> `^6.1.0`).
- The redirect route (`/:shortUrl([a-zA-Z0-9-]+).?`, the single entry point for all go.gov.sg redirect traffic) used a path-to-regexp v6 syntax (inline per-param regex + bare `?` optional suffix) that path-to-regexp v8 (bundled with Express 5) no longer supports at all. Replaced with a plain `/:shortUrl` route + a new `shortUrlRouteGuard` middleware that reproduces the same accept/reject behavior in application code.
- `express-joi-validation@^5.0.1` unconditionally does `req.query = ...` on every `validator.query(...)`-validated request (6 routes: link-statistics, external-v1, link-audit, user, directory, qrcode). Express 5's `req.query` is getter-only, so this **throws on every request** to those routes. Confirmed empirically with a red/green test before bumping to `^6.1.0`, which added the writability check needed to support Express 5.
- Found and fixed one regression during final review: Express 5's router decodes the matched `:shortUrl` path segment before any route handler runs, so malformed percent-encoding (e.g. `/%`, `/%zz`) was turning into a 500 instead of a 404. Fixed in the app's error handler, with regression tests.
- Everything else on the standard Express 5 breaking-changes list was checked and found not to apply to this codebase (see the "Global Constraints" section of the plan doc, kept locally, not tracked in this branch per this repo's existing convention).

## How this was built

Planned and executed via Claude Code's subagent-driven-development workflow: a plan with 3 tasks, each implemented and reviewed independently, followed by a whole-branch review (which caught the percent-encoding regression above) and a fix round.

## Known gap — needs confirmation in CI, not locally verified

Local Docker (Colima) was in a broken, VM-level I/O-error state in the sandbox this was built in, unrelated to this branch. As a result, the following were **not** run locally and should be treated as unconfirmed until this PR's CI jobs pass:

- Integration tests (`pnpm run test:integration`) — the `express-joi-validation` fix has route-level coverage in the unit suite via `test/server/api/{qrcodeRoute,LinkStatsRoute,LoginRoute,LogoutRoute,LinksRoute}.test.ts` (real Express 5 app + supertest, not mocks), but not the full DB/Redis-backed integration suite.
- End-to-end Playwright tests.
- A manual curl smoke test of the redirect route against a live server.

Locally verified: `pnpm run lint`, `pnpm run format:check`, `pnpm typecheck`, `pnpm run build`, and the full `pnpm test` unit suite (61 suites / 499 passing / 4 pre-existing skips) all pass clean.

## Test plan

- [ ] CI `Integration Tests` job passes
- [ ] CI `End To End Tests` job passes (all 3 browsers)
- [ ] CI `Test`, `Typecheck`, `Lint`, `Format`, `Build` jobs pass
- [x] Full unit suite passes locally (61/61 suites, 499/499 non-skipped tests)
- [x] `express-joi-validation` query-crash fix empirically proven via red/green test (`src/server/api/__tests__/expressJoiValidationQuerySupport.test.ts`)
- [x] Redirect route accept/reject behavior (including the trailing-dot and percent-encoding edge cases) covered by `src/server/modules/redirect/__tests__/shortUrlRouteGuard.test.ts`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Greptile Summary</h3></summary>

The PR upgrades Express and its validation/types dependencies to Express 5-compatible versions while preserving public short-link routing and malformed-URI handling.
- Replaces the unsupported inline route expression with a `/:shortUrl` route and application-level guard.
- Adds Express 5 query-validation compatibility coverage.
- Routes malformed API parameters to JSON errors and malformed redirect parameters to the short-link 404 page.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

The PR is not yet safe to merge because mixed-case API paths can still receive redirect-specific HTML when malformed path parameters trigger URI decoding errors.

Express matches the `/api` mount case-insensitively, while `handleUriError` classifies requests using a case-sensitive lowercase prefix check, so requests such as `PATCH /API/v1/urls/%zz` remain outside the intended JSON error branch.

**Files Needing Attention:** src/server/util/handleUriError.ts and src/server/util/__tests__/handleUriError.test.ts
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| package.json | Upgrades Express, Express types, and express-joi-validation to Express 5-compatible versions. |
| src/server/index.ts | Replaces the unsupported redirect route expression, installs the route guard, and delegates URI decoding errors to a dedicated handler. |
| src/server/modules/redirect/shortUrlRouteGuard.ts | Reproduces the prior short-link character and optional trailing-dot constraints in middleware. |
| src/server/util/handleUriError.ts | Separates malformed API and redirect responses, but its case-sensitive API-prefix check leaves the previously reported mixed-case behavior outstanding. |
| src/server/util/__tests__/handleUriError.test.ts | Covers lowercase API and redirect URI errors but does not cover the outstanding mixed-case API path. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  A[Incoming request] --> B{API route?}
  B -->|Yes| C[API router]
  B -->|No| D[Short-link route]
  C --> E{Malformed encoded parameter?}
  D --> F{Parameter decodes?}
  E -->|Yes| G[Global URIError handler]
  F -->|No| G
  F -->|Yes| H[Short URL guard]
  G --> I{Path starts with lowercase /api/?}
  I -->|Yes| J[JSON 400]
  I -->|No| K[Redirect HTML 404]
  H -->|Valid| L[Redirect controller]
  H -->|Invalid| K
```
</details>

<sub>Reviews (6): Last reviewed commit: ["style: fix formatting in handleUriError ..."](https://github.com/opengovsg/gogovsg/commit/b746d8ac312b07f02bf6f9d57cb26d1e49fedcd6) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=52319510)</sub>

**Context used:**

- Knowledge Base — [Public link redirection](https://app.greptile.com/opengovsg/-/custom-context/knowledge-base/opengovsg/gogovsg/-/docs/public-link-redirection.md)
- Knowledge Base — [Link APIs](https://app.greptile.com/opengovsg/-/custom-context/knowledge-base/opengovsg/gogovsg/-/docs/link-apis.md)

### CODE REVIEW

File: src/server/index.ts [hunk 1]
```diff
@@ -247,6 +258,24 @@ initDb()
...
+        // Express 5's router decodes the matched `:shortUrl` path segment
+        // (via decodeURIComponent) before any route handler runs, and
+        // throws a URIError straight into this error handler when the
+        // segment is malformed percent-encoding (e.g. `/%`, `/%zz`,
+        // `/%c0%af`). That's not a server error — it just means the
+        // request doesn't resolve to a short URL, same as any other
+        // non-matching path, so render the same 404 the catch-all handler
+        // above renders instead of the generic 500 page.
+        const shortUrl = req.path.slice(1)
+        res.status(404).render(ERROR_404_PATH, {
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply src/server/index.ts]: @cursoragent verify if valid?

[the remark this answers is not in the document]
Comment (adriangohjw) [reply src/server/index.ts]: @cursoragent add fix + add appropriate tests

**Comment by adriangohjw** [comment]
@cursoragent fix formattng CI error

**Review by seaerchin — COMMENTED** [review]
## Automated review — focus: breaking changes & compatibility

**Verdict: LOW RISK** — this is exemplary breaking-change handling, with one real caveat on test coverage.

This PR catches and fixes three distinct Express 5 breaking changes, not just a version bump:
1. The redirect route (`/:shortUrl([a-zA-Z0-9-]+).?` — the single entry point for all go.gov.sg redirect traffic) used path-to-regexp v6 syntax (inline per-param regex + bare `?`) that path-to-regexp v8 (bundled with Express 5) doesn't support **at all**. Replaced with a plain `/:shortUrl` route plus a new `shortUrlRouteGuard` middleware reproducing the same accept/reject behavior in application code.
2. `express-joi-validation@^5.0.1` unconditionally reassigns `req.query`, which is getter-only in Express 5 — this would throw on every request to 6 validated routes. Caught with an empirical red/green test before bumping to `^6.1.0`, which adds the writability check Express 5 needs.
3. Found during final review (not by luck — an explicit whole-branch review step): Express 5's router now decodes the `:shortUrl` path segment before any handler runs, so malformed percent-encoding (`/%`, `/%zz`) was producing a 500 instead of a 404. Fixed with regression tests.

- **Real gap, self-disclosed:** local Docker was broken in the sandbox this was built in, so **integration tests and Playwright e2e were not run locally** — only the unit suite (61/61 suites, 499/499 tests) and route-level supertest coverage against a real Express 5 app. Treat this as unconfirmed until CI's integration and e2e jobs actually pass green on this PR.
- Overlaps on `express` version with several dependabot PRs (#2514, #2512, #2509, #2461, #2456) that bump express within the 4.x line — those become moot/need rebasing once this lands.

**Before merging:** do not merge until CI's Integration and End-to-End jobs are confirmed green — the author was explicit that those weren't exercised locally.

*Posted by an automated review pass focused on breaking changes/compatibility.*

**Review by seaerchin — COMMENTED** [review]
express 5 passes the error to the callback function, so we need to throw the error upwards

**Review by seaerchin — APPROVED** [review]
